operations/ntl/cc/kernels/ntl_matrix_ops.cc

```diff
@@ -0,0 +1,193 @@
+    auto rows = in->dim_size(0), cols = in->dim_size(1);
+    NTLMatrix m(rows, cols);
+
+    auto mat = in->matrix<string>();
+
+    for(int i = 0; i < rows; i++) {
+        for(int j = 0; j < cols; j++) {
+            m.m[i][j] = NTL::conv<NTL::ZZ_p>(mat(i, j).c_str());
+        }
+    }
+
+    out->scalar<Variant>()() = std::move(m);
+  }
+};
+
+template<typename T>
+struct NTLToNativeOpFunctor {
+  void operator()(const NTLMatrix* in, Tensor* out) {
+    auto rows = in->m.NumRows(), cols = in->m.NumCols();
+
+    for(int i = 0; i < rows; i++) {
+       for(int j = 0; j < cols; j++) {
+         out->matrix<T>()(i, j) = (T)NTL::conv<long>(in->m[i][j]);
+       }
+     }
+  }
+};
+
+template<>
+struct NTLToNativeOpFunctor<string> {
+  void operator()(const NTLMatrix* in, Tensor* out) {
+    auto rows = in->m.NumRows(), cols = in->m.NumCols();
+
+    for(int i = 0; i < rows; i++) {
+       for(int j = 0; j < cols; j++) {
+         std::stringstream buffer;
+         buffer << in->m[i][j];
+         out->matrix<string>()(i, j) = buffer.str();
+       }
+     }
+  }
+};
+}
+
+template <typename T>
+class CreateNTLMatrixOp : public OpKernel {
+public:
+  explicit CreateNTLMatrixOp(OpKernelConstruction* context) : OpKernel(context) {}
+
+  void Compute(OpKernelContext* ctx) override {
+    const Tensor& value = ctx->input(0);
+    const Tensor& mod = ctx->input(1);
+
+    // TODO support vector?? scalar??
+    OP_REQUIRES(
+        ctx, TensorShapeUtils::IsMatrix(value.shape()),
+        errors::InvalidArgument(
+            "value expected to be a 2d matrix ",
+            "but got shape: ", value.shape().DebugString()));
+
+    OP_REQUIRES(
+        ctx, TensorShapeUtils::IsScalar(mod.shape()),
+        errors::InvalidArgument(
+            "value expected to be a scalar ",
+            "but got shape: ", mod.shape().DebugString()));
+
+    NTL::ZZ_p::init(NTL::ZZ(mod.scalar<int64>()()));
+
+    Tensor* result;
+    OP_REQUIRES_OK(ctx, ctx->allocate_output(0, TensorShape{}, &result));
+
+    functor::CreateNTLMatrixOpFunctor<T>()(&value, result);
+  }
+};
+
+Status GetNTLMatrix(OpKernelContext* ctx, int index, const NTLMatrix** wrapper) {
+    const Tensor& input = ctx->input(index);
+
+    // TODO: check scalar type
+    const NTLMatrix * w = input.scalar<Variant>()().get<NTLMatrix>();
+    if(w == nullptr) {
+        return errors::InvalidArgument(
+        "Input handle is not a ntl wrapper. Saw: '",
+        input.scalar<Variant>()().DebugString(), "'");
+    }
+
+    *wrapper = w;
+    return Status::OK();
+}
+
+class MatMulNTLOp : public OpKernel {
+public:
+  explicit MatMulNTLOp(OpKernelConstruction* context) : OpKernel(context) {}
+
+  void Compute(OpKernelContext* ctx) override {
+    const NTLMatrix * m1 = nullptr;
+    OP_REQUIRES_OK(ctx, GetNTLMatrix(ctx, 0, &m1));
+
+    const NTLMatrix * m2 = nullptr;
+    OP_REQUIRES_OK(ctx, GetNTLMatrix(ctx, 1, &m2));
+
+    const Tensor& mod = ctx->input(2);
```

Comment:
I'm wondering if this shouldn't be taken from one of the other two tensors instead of being given explicitly. If explicit then I see users grapping it from one of those anyways, and it would be a fault to pass in anything else.

Reply:
Yeah good idea.